extensions/src/AWSSDK.Extensions.Bedrock.MEAI/BedrockChatClient.cs

```diff
@@ -0,0 +1,645 @@
+                    choice = !string.IsNullOrWhiteSpace(r.RequiredFunctionName) ?
+                        new ToolChoice() { Tool = new() { Name = r.RequiredFunctionName } } :
+                        new ToolChoice() { Any = new() };
+                    break;
+            }
+
+            return new()
+            {
+                ToolChoice = choice,
+                Tools = tools,
+            };
+        }
+
+        return null;
+    }
+
+    /// <summary>Creates an <see cref="InferenceConfiguration"/> from the specified options.</summary>
+    private static InferenceConfiguration CreateInferenceConfiguration(ChatOptions? options) =>
+        new()
+        {
+            MaxTokens = options?.MaxOutputTokens,
+            StopSequences = options?.StopSequences?.ToList(),
+            Temperature = options?.Temperature,
+            TopP = options?.TopP,
+        };
+
+    /// <summary>Creates a <see cref="Document"/> from the specified options to use as the additional model request options.</summary>
+    private static Document CreateAdditionalModelRequestFields(ChatOptions? options)
+    {
+        Document d = default;
+
+        if (options is not null)
+        {
+            if (options.TopK is int topK)
+            {
+                d.Add("k", topK);
+            }
+
+            if (options.FrequencyPenalty is float frequencyPenalty)
+            {
+                d.Add("frequency_penalty", frequencyPenalty);
+            }
+
+            if (options.PresencePenalty is float presencePenalty)
+            {
+                d.Add("presence_penalty", presencePenalty);
+            }
+
+            if (options.Seed is long seed)
+            {
+                d.Add("seed", seed);
+            }
+
+            if (options.AdditionalProperties is { } props)
+            {
+                foreach (KeyValuePair<string, object?> prop in props)
+                {
+                    switch (prop.Value)
+                    {
+                        case bool propBool: d.Add(prop.Key, propBool); break;
+                        case int propInt32: d.Add(prop.Key, propInt32); break;
+                        case long propInt64: d.Add(prop.Key, propInt64); break;
+                        case float propSingle: d.Add(prop.Key, propSingle); break;
+                        case double propDouble: d.Add(prop.Key, propDouble); break;
+                        case string propString: d.Add(prop.Key, propString); break;
+                        case null: d.Add(prop.Key, default); break;
+                        case JsonElement json: d.Add(prop.Key, ToDocument(json)); break;
+                        default:
+                            try
+                            {
+                                d.Add(prop.Key, ToDocument(JsonSerializer.SerializeToElement(prop.Value, BedrockJsonContext.DefaultOptions.GetTypeInfo(prop.Value.GetType()))));
+                            }
+                            catch { }
```

Comment:
I'll update it to log the failure.

> There might be other areas in this class that are worth logging as well.

Most logging with Microsoft.Extensions.AI is designed to be external to the leaf client, so that the leaf client implementation (like this one) needn't concern itself with such cross-cutting details. Someone can add logging of all request/responses, for example, by using a delegating logging chat client as part of their pipeline, e.g.
```C#
.AddChatClient(bedrock.AsChatClient())
    .UseFunctionInvocation() // logs details about function invocations
    .UseLogging() // logs all requests/responses
    ... // etc
    .Build(services) // all of the above will query these services for a logger factory to use
```

Reply:
I'm hoping as part of V4 I can squeeze revamping the SDK's logging so we have a way to integrate with Microsoft.Extensions.Logging. The internal API that used won't change but how you turn on SDK logging will have some changes. I just need to find the time to finish the my in progress work 😄 .